net/message.go

```diff
@@ -0,0 +1,95 @@
+package net
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/athanorlabs/go-relayer/common"
+)
+
+// Message must be implemented by all network messages
+type Message interface {
+	Type() byte
+}
+
+func decodeMessage(b []byte) (Message, error) {
+	// 1-byte type followed by at least 2-bytes of JSON (`{}`)
+	if len(b) < 3 {
+		return nil, errors.New("invalid message bytes")
+	}
+
+	msgType := b[0]
+	msgJSON := b[1:]
+	var msg Message
+
+	switch msgType {
+	case TransactionResponseType:
+		msg = &TransactionResponse{}
+	case TransactionRequestType:
+		msg = &TransactionRequest{}
+	default:
+		return nil, fmt.Errorf("invalid message type %d", msgType)
+	}
+
+	if err := json.Unmarshal(msgJSON, &msg); err != nil {
+		return nil, fmt.Errorf("failed to decode message with type %d: %w", msg.Type(), err)
+	}
+	return msg, nil
+}
+
+const (
+	TransactionRequestType byte = iota
+	TransactionResponseType
+)
+
+type TransactionRequest struct {
+	*common.SubmitTransactionRequest
+}
```

Comment:
I'm not familiar enough with the code yet to say what is ideal, but you have a few other options:
```
 // inherit all functions, can't override locally
type TransactionRequest = common.SubmitTransactionRequest
```
```
// new type, no functions inherited
type TransactionRequest common.SubmitTransactionRequest
```

You could leave it the way it is, but remove the pointer, as this is the only side that ever creates the inner object. You'll still be able to send the inner value using `tr.SubmitTransactionRequest`.

Reply:
I'll just remove the pointer, since I need additional functions on the type